[bug][tests] CI can't collect support bundle from standalone clusters

Faced on: https://github.com/k0rdent/kcm/actions/runs/16595422148/job/46941512751?pr=1792

```
  STEP: collecting the support bundle from the ci-90633-aws-0 cluster @ 07/29/25 12:40:07.781
  running: make KUBECONFIG=/home/runner/work/kcm/kcm/ci-90633-aws-0-kubeconfig support-bundle
  make[1]: Entering directory '/home/runner/work/kcm/kcm'

  ============ Collectors summary =============
  Succeeded (S), eXcluded (X), Failed (F)
  =============================================
  No collectors executed

  ============ Redactors summary =============
  No redactors executed

  ============= Analyzers summary =============
  Succeeded (S), eXcluded (X), Failed (F)
  =============================================
  No analyzers executed

  Duration: 0ms

  Error: failed to run collect and analyze process: failed to generate support bundle: failed to run collectors: failed to check RBAC for collectors: failed to run subject review: Post "http://localhost:8080/apis/authorization.k8s.io/v1/selfsubjectaccessreviews": dial tcp [::1]:8080: connect: connection refused
  make[1]: *** [Makefile:454: support-bundle] Error 1
  make[1]: Leaving directory '/home/runner/work/kcm/kcm'
  Warning: failed to collect the support bundle: make KUBECONFIG=/home/runner/work/kcm/kcm/ci-90633-aws-0-kubeconfig support-bundle failed with error: (exit status 2): 
```

The support bundle from standalone cluster deployments will be extremely helpful to debug hosted cluster deployment issues.

--- a/test/e2e/clusterdeployment/azure/azure.go
+++ b/test/e2e/clusterdeployment/azure/azure.go
@@ -47,14 +47,14 @@ func CheckEnv() {
 func PopulateEnvVars(architecture config.Architecture) {
 	GinkgoHelper()
 
+	GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageGallery, "aksubuntu-38d80f77-467a-481f-a8d4-09b6d4220bd2")
+
 	switch architecture {
 	case config.ArchitectureAmd64:
-		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageGallery, "aksazurelinux-f7c7cda5-1c9a-4bdc-a222-9614c968580b")
-		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageName, "V2")
-		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageVersion, "202409.23.0")
+		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageName, "2204containerd")
+		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageVersion, "202508.06.0")
 		GinkgoT().Setenv(clusterdeployment.EnvVarAzureVMSize, "Standard_A4_v2")
 	case config.ArchitectureArm64:
-		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageGallery, "aksubuntu-38d80f77-467a-481f-a8d4-09b6d4220bd2")
 		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageName, "2204gen2arm64containerd")
 		GinkgoT().Setenv(clusterdeployment.EnvVarAzureImageVersion, "202507.29.0")
 		GinkgoT().Setenv(clusterdeployment.EnvVarAzureVMSize, "Standard_D4plds_v5")

--- a/test/e2e/config/config.go
+++ b/test/e2e/config/config.go
@@ -82,21 +82,22 @@ type ClusterTestingConfig struct {
 }
 
 func Parse() error {
-	if len(configBytes) == 0 {
-		Initialize()
-		return nil
-	}
 	parseOnce.Do(func() {
-		err := yaml.Unmarshal(configBytes, &Config)
-		if err != nil {
+		if len(configBytes) == 0 {
+			initialize()
+			return
+		}
+
+		if err := yaml.Unmarshal(configBytes, &Config); err != nil {
 			errParse = fmt.Errorf("failed to decode base64 configuration: %w", err)
 			return
 		}
+		applyDefaultConfiguration()
 	})
 	return errParse
 }
 
-func Initialize() {
+func initialize() {
 	providers := []TestingProvider{
 		TestingProviderAWS,
 		TestingProviderAzure,
@@ -109,6 +110,14 @@ func Initialize() {
 	Config = make(map[TestingProvider][]ProviderTestingConfig)
 	for _, provider := range providers {
 		Config[provider] = getDefaultTestingConfiguration()
+	}
+}
+
+func applyDefaultConfiguration() {
+	for provider, configs := range Config {
+		if len(configs) == 0 {
+			Config[provider] = getDefaultTestingConfiguration()
+		}
 	}
 }
 

--- a/test/e2e/e2e_suite_test.go
+++ b/test/e2e/e2e_suite_test.go
@@ -41,7 +41,11 @@ import (
 	"github.com/K0rdent/kcm/test/utils"
 )
 
-var clusterTemplates []string
+var (
+	clusterTemplates []string
+
+	kc *kubeclient.KubeClient
+)
 
 // Run e2e tests using the Ginkgo runner.
 func TestE2E(t *testing.T) {
@@ -69,7 +73,7 @@ var _ = BeforeSuite(func() {
 		Expect(err).NotTo(HaveOccurred())
 	}
 
-	kc := kubeclient.NewFromLocal(internalutils.DefaultSystemNamespace)
+	kc = kubeclient.NewFromLocal(internalutils.DefaultSystemNamespace)
 
 	// TODO: remove after https://github.com/k0rdent/kcm/issues/1575 is fixed
 	removeInfobloxProvider(kc.CrClient)
@@ -102,7 +106,7 @@ var _ = BeforeSuite(func() {
 var _ = AfterSuite(func() {
 	if cleanup() {
 		By("collecting the support bundle from the management cluster")
-		logs.SupportBundle("")
+		logs.SupportBundle(kc, "")
 
 		By("removing the controller-manager")
 		cmd := exec.Command("make", "dev-destroy")

--- a/test/e2e/kubeclient/kubeclient.go
+++ b/test/e2e/kubeclient/kubeclient.go
@@ -41,7 +41,6 @@ import (
 
 	kcmv1 "github.com/K0rdent/kcm/api/v1beta1"
 	"github.com/K0rdent/kcm/internal/utils/status"
-	"github.com/K0rdent/kcm/test/e2e/logs"
 	"github.com/K0rdent/kcm/test/scheme"
 )
 
@@ -66,47 +65,59 @@ func NewFromLocal(namespace string) *KubeClient {
 // the kubeconfig from secret it needs an existing kubeclient.
 func (kc *KubeClient) NewFromCluster(ctx context.Context, namespace, clusterName string) *KubeClient {
 	GinkgoHelper()
-	return newKubeClient(kc.GetKubeconfigSecretData(ctx, clusterName), namespace)
+
+	secretData, err := kc.GetKubeconfigSecretData(ctx, clusterName)
+	Expect(err).To(Succeed())
+	return newKubeClient(secretData, namespace)
 }
 
 // WriteKubeconfig writes the kubeconfig for the given clusterName to the
 // test/e2e directory returning the path to the file and a function to delete
 // it later.
-func (kc *KubeClient) WriteKubeconfig(ctx context.Context, clusterName string) (path, b64Raw string, cleanup func() error) {
+func (kc *KubeClient) WriteKubeconfig(ctx context.Context, clusterName string) (path, b64Raw string, cleanup func() error, err error) {
 	GinkgoHelper()
 
-	secretData := kc.GetKubeconfigSecretData(ctx, clusterName)
+	secretData, err := kc.GetKubeconfigSecretData(ctx, clusterName)
+	if err != nil {
+		return "", "", nil, fmt.Errorf("failed to get kubeconfig secret data: %w", err)
+	}
 
 	kcFile, err := os.CreateTemp("", clusterName+"-kubeconfig")
-	Expect(err).To(Succeed())
+	if err != nil {
+		return "", "", nil, err
+	}
 	defer kcFile.Close() //nolint:errcheck // no need
 
 	_, err = kcFile.Write(secretData)
-	Expect(err).To(Succeed())
-
-	logs.Println("Successfully created kubeconfig " + kcFile.Name())
+	if err != nil {
+		return "", "", nil, err
+	}
 
 	cleanup = func() error {
 		if err := os.Remove(kcFile.Name()); err != nil && !errors.Is(err, fs.ErrNotExist) {
 			return err
 		}
-
 		return nil
 	}
 
-	return kcFile.Name(), base64.StdEncoding.EncodeToString(bytes.TrimSpace(secretData)), cleanup
+	return kcFile.Name(), base64.StdEncoding.EncodeToString(bytes.TrimSpace(secretData)), cleanup, nil
 }
 
-func (kc *KubeClient) GetKubeconfigSecretData(ctx context.Context, clusterName string) []byte {
+func (kc *KubeClient) GetKubeconfigSecretData(ctx context.Context, clusterName string) ([]byte, error) {
 	GinkgoHelper()
 
-	secret, err := kc.Client.CoreV1().Secrets(kc.Namespace).Get(ctx, clusterName+"-kubeconfig", metav1.GetOptions{})
+	kubeconfigSecretName := clusterName + "-kubeconfig"
+	secret, err := kc.Client.CoreV1().Secrets(kc.Namespace).Get(ctx, kubeconfigSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster: %q kubeconfig secret: %w", clusterName, err)
+	}
 	Expect(err).NotTo(HaveOccurred(), "failed to get cluster: %q kubeconfig secret", clusterName)
 
 	secretData, ok := secret.Data["value"]
-	Expect(ok).To(BeTrue(), "kubeconfig secret %q has no 'value' key", clusterName)
-
-	return secretData
+	if !ok {
+		return nil, fmt.Errorf("kubeconfig secret %q has no 'value' key", kubeconfigSecretName)
+	}
+	return secretData, nil
 }
 
 // getLocalKubeConfig returns the kubeconfig file content.

--- a/test/e2e/logs/logs.go
+++ b/test/e2e/logs/logs.go
@@ -15,33 +15,44 @@
 package logs
 
 import (
+	"context"
 	"fmt"
-	"os"
 	"os/exec"
-	"path/filepath"
 	"time"
 
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
 
+	"github.com/K0rdent/kcm/test/e2e/kubeclient"
 	"github.com/K0rdent/kcm/test/utils"
 )
 
 // SupportBundle collects the support bundle from the specified cluster.
 // If the clusterName is unset, it collects the support bundle from the management cluster.
-func SupportBundle(clusterName string) {
-	var args []string
+func SupportBundle(kc *kubeclient.KubeClient, clusterName string) {
+	var (
+		args        []string
+		kubeCfgPath string
+		cleanupFunc func() error
+		err         error
+	)
 	if clusterName != "" {
-		dir, err := os.Getwd()
-		Expect(err).NotTo(HaveOccurred())
+		kubeCfgPath, _, cleanupFunc, err = kc.WriteKubeconfig(context.Background(), clusterName)
+		if err != nil {
+			utils.WarnError(fmt.Errorf("failed to write %s cluster kubeconfig: %w", clusterName, err))
+			return
+		}
 
-		args = append(args, fmt.Sprintf("KUBECONFIG=%s", filepath.Join(dir, clusterName+"-kubeconfig")))
+		args = append(args, fmt.Sprintf("KUBECONFIG=%s", kubeCfgPath))
 	}
 	args = append(args, "support-bundle")
 	cmd := exec.Command("make", args...)
-	_, err := utils.Run(cmd)
-	if err != nil {
+	if _, err = utils.Run(cmd); err != nil {
 		utils.WarnError(fmt.Errorf("failed to collect the support bundle: %w", err))
+	}
+	if cleanupFunc != nil {
+		err = cleanupFunc()
+		Expect(err).NotTo(HaveOccurred())
 	}
 }
 

--- a/test/e2e/multi_provider_test.go
+++ b/test/e2e/multi_provider_test.go
@@ -16,6 +16,7 @@ package e2e
 
 import (
 	"context"
+	"fmt"
 	"time"
 
 	sourcev1 "github.com/fluxcd/source-controller/api/v1"
@@ -94,7 +95,12 @@ var _ = Context("Multi Cloud Templates", Label("provider:multi-cloud", "provider
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
+
+			for _, clusterName := range []string{awsClusterDeploymentName, azureClusterDeploymentName} {
+				By(fmt.Sprintf("collecting the support bundle from the %s cluster", awsClusterDeploymentName))
+				logs.SupportBundle(kc, clusterName)
+			}
 		}
 
 		By("deleting resources")

--- a/test/e2e/provider_adopted_test.go
+++ b/test/e2e/provider_adopted_test.go
@@ -88,7 +88,7 @@ var _ = Describe("Adopted Cluster Templates", Label("provider:cloud", "provider:
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("Collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 		}
 
 		if cleanup() {
@@ -164,7 +164,9 @@ var _ = Describe("Adopted Cluster Templates", Label("provider:cloud", "provider:
 
 			// create the adopted cluster using the AWS standalone cluster
 			var rawData string
-			_, rawData, kubecfgDeleteFunc = kc.WriteKubeconfig(context.Background(), clusterName)
+			var err error
+			_, rawData, kubecfgDeleteFunc, err = kc.WriteKubeconfig(context.Background(), clusterName)
+			Expect(err).To(Succeed())
 			Expect(os.Setenv(clusterdeployment.EnvVarAdoptedKubeconfigData, rawData)).Should(Succeed())
 			credential.Apply("", "adopted")
 

--- a/test/e2e/provider_aws_test.go
+++ b/test/e2e/provider_aws_test.go
@@ -90,11 +90,11 @@ var _ = Describe("AWS Templates", Label("provider:cloud", "provider:aws"), Order
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 
 			for _, clusterName := range standaloneClusters {
 				By(fmt.Sprintf("collecting the support bundle from the %s cluster", clusterName))
-				logs.SupportBundle(clusterName)
+				logs.SupportBundle(kc, clusterName)
 			}
 		}
 
@@ -211,12 +211,13 @@ var _ = Describe("AWS Templates", Label("provider:cloud", "provider:aws"), Order
 				// TODO(#472): Ideally we shouldn't use Make here and should just
 				// convert these Make targets into Go code, but this will require a
 				// helmclient.
-				kubeCfgPath, _, kubecfgDeleteFunc := kc.WriteKubeconfig(context.Background(), sdName)
+				kubeCfgPath, _, kubecfgDeleteFunc, err := kc.WriteKubeconfig(context.Background(), sdName)
+				Expect(err).To(Succeed())
 				kubeconfigDeleteFuncs = append(kubeconfigDeleteFuncs, kubecfgDeleteFunc)
 
 				GinkgoT().Setenv("KUBECONFIG", kubeCfgPath)
 				cmd := exec.Command("make", "test-apply")
-				_, err := utils.Run(cmd)
+				_, err = utils.Run(cmd)
 				Expect(err).NotTo(HaveOccurred())
 				Expect(os.Unsetenv("KUBECONFIG")).To(Succeed())
 

--- a/test/e2e/provider_azure_test.go
+++ b/test/e2e/provider_azure_test.go
@@ -75,11 +75,11 @@ var _ = Context("Azure Templates", Label("provider:cloud", "provider:azure"), Or
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 
 			for _, clusterName := range standaloneClusters {
 				By(fmt.Sprintf("collecting the support bundle from the %s cluster", clusterName))
-				logs.SupportBundle(clusterName)
+				logs.SupportBundle(kc, clusterName)
 			}
 		}
 
@@ -160,13 +160,14 @@ var _ = Context("Azure Templates", Label("provider:cloud", "provider:azure"), Or
 				// setup environment variables for deploying the hosted template (subnet name, etc)
 				azure.SetAzureEnvironmentVariables(sdName, kc)
 
-				kubeCfgPath, _, kubecfgDeleteFunc := kc.WriteKubeconfig(context.Background(), sdName)
+				kubeCfgPath, _, kubecfgDeleteFunc, err := kc.WriteKubeconfig(context.Background(), sdName)
+				Expect(err).To(Succeed())
 				kubeconfigDeleteFuncs = append(kubeconfigDeleteFuncs, kubecfgDeleteFunc)
 
 				By("Deploy onto standalone cluster")
 				GinkgoT().Setenv("KUBECONFIG", kubeCfgPath)
 				cmd := exec.Command("make", "test-apply")
-				_, err := utils.Run(cmd)
+				_, err = utils.Run(cmd)
 				Expect(err).NotTo(HaveOccurred())
 				Expect(os.Unsetenv("KUBECONFIG")).To(Succeed())
 

--- a/test/e2e/provider_azure_test.go
+++ b/test/e2e/provider_azure_test.go
@@ -19,7 +19,6 @@ import (
 	"fmt"
 	"os"
 	"os/exec"
-	"slices"
 	"time"
 
 	. "github.com/onsi/ginkgo/v2"
@@ -47,28 +46,11 @@ var _ = Context("Azure Templates", Label("provider:cloud", "provider:azure"), Or
 	)
 
 	BeforeAll(func() {
-		By("Get testing configuration")
-		providerConfigs := config.Config[config.TestingProviderAzure]
-
 		By("Ensuring that env vars are set correctly")
 		azure.CheckEnv()
 
 		By("Creating kube client")
 		kc = kubeclient.NewFromLocal(internalutils.DefaultSystemNamespace)
-
-		By("Providing cluster identity and credentials")
-		if slices.ContainsFunc(providerConfigs, func(providerConfig config.ProviderTestingConfig) bool {
-			return templates.GetType(providerConfig.Template) == templates.TemplateAzureAKS
-		}) {
-			credential.Apply("", "aks")
-		}
-
-		if slices.ContainsFunc(providerConfigs, func(providerConfig config.ProviderTestingConfig) bool {
-			return templates.GetType(providerConfig.Template) == templates.TemplateAzureStandaloneCP ||
-				templates.GetType(providerConfig.Template) == templates.TemplateAzureHostedCP
-		}) {
-			credential.Apply("", "azure")
-		}
 	})
 
 	AfterAll(func() {
@@ -105,6 +87,13 @@ var _ = Context("Azure Templates", Label("provider:cloud", "provider:azure"), Or
 			sdName := clusterdeployment.GenerateClusterName(fmt.Sprintf("azure-%d", i))
 			sdTemplate := testingConfig.Template
 			sdTemplateType := templates.GetType(sdTemplate)
+
+			By("Providing cluster identity and credentials")
+			provider := "azure"
+			if sdTemplateType == templates.TemplateAzureAKS {
+				provider = "aks"
+			}
+			credential.Apply("", provider)
 
 			// Supported architectures for Azure standalone deployment: amd64, arm64
 			Expect(testingConfig.Architecture).To(SatisfyAny(

--- a/test/e2e/provider_gcp_test.go
+++ b/test/e2e/provider_gcp_test.go
@@ -60,11 +60,11 @@ var _ = Context("GCP Templates", Label("provider:cloud", "provider:gcp"), Ordere
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 
 			for _, clusterName := range standaloneClusters {
 				By(fmt.Sprintf("collecting the support bundle from the %s cluster", clusterName))
-				logs.SupportBundle(clusterName)
+				logs.SupportBundle(kc, clusterName)
 			}
 		}
 
@@ -148,13 +148,14 @@ var _ = Context("GCP Templates", Label("provider:cloud", "provider:gcp"), Ordere
 			standaloneClient := new(kubeclient.KubeClient)
 			var hdName string
 			if testingConfig.Hosted != nil {
-				kubeCfgPath, _, kubecfgDeleteFunc := kc.WriteKubeconfig(context.Background(), sdName)
+				kubeCfgPath, _, kubecfgDeleteFunc, err := kc.WriteKubeconfig(context.Background(), sdName)
+				Expect(err).To(Succeed())
 				kubeconfigDeleteFuncs = append(kubeconfigDeleteFuncs, kubecfgDeleteFunc)
 
 				By("Deploy onto standalone cluster")
 				GinkgoT().Setenv("KUBECONFIG", kubeCfgPath)
 				cmd := exec.Command("make", "test-apply")
-				_, err := utils.Run(cmd)
+				_, err = utils.Run(cmd)
 				Expect(err).NotTo(HaveOccurred())
 				Expect(os.Unsetenv("KUBECONFIG")).To(Succeed())
 

--- a/test/e2e/provider_remote_test.go
+++ b/test/e2e/provider_remote_test.go
@@ -78,7 +78,7 @@ var _ = Describe("Remote Cluster Templates", Label("provider:cloud", "provider:r
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 		}
 
 		if cleanup() {

--- a/test/e2e/provider_vsphere_test.go
+++ b/test/e2e/provider_vsphere_test.go
@@ -61,11 +61,11 @@ var _ = Context("vSphere Templates", Label("provider:onprem", "provider:vsphere"
 		// If we failed collect the support bundle before the cleanup
 		if CurrentSpecReport().Failed() && cleanup() {
 			By("Collecting the support bundle from the management cluster")
-			logs.SupportBundle("")
+			logs.SupportBundle(kc, "")
 
 			for _, clusterName := range standaloneClusterNames {
 				By(fmt.Sprintf("Collecting the support bundle from the %s cluster", clusterName))
-				logs.SupportBundle(clusterName)
+				logs.SupportBundle(kc, clusterName)
 			}
 		}
 
@@ -146,12 +146,13 @@ var _ = Context("vSphere Templates", Label("provider:onprem", "provider:vsphere"
 			standaloneClient := new(kubeclient.KubeClient)
 			var hdName string
 			if testingConfig.Hosted != nil {
-				kubeCfgPath, _, kubecfgDeleteFunc := kc.WriteKubeconfig(context.Background(), sdName)
+				kubeCfgPath, _, kubecfgDeleteFunc, err := kc.WriteKubeconfig(context.Background(), sdName)
+				Expect(err).To(Succeed())
 				kubeconfigDeleteFuncs = append(kubeconfigDeleteFuncs, kubecfgDeleteFunc)
 
 				By("Deploy onto standalone cluster")
 				GinkgoT().Setenv("KUBECONFIG", kubeCfgPath)
-				_, err := utils.Run(exec.Command("make", "test-apply"))
+				_, err = utils.Run(exec.Command("make", "test-apply"))
 				Expect(err).NotTo(HaveOccurred())
 				Expect(os.Unsetenv("KUBECONFIG")).To(Succeed())
 

